File: jpp/lab/l4/z1/DHSetup.cpp
#include <random>
#include "GF.cpp"


template <typename T>
class DHSetup {
private:
    T generator;
public:
    DHSetup(T element) : generator(element){
        generator = ~element; /// random generator from element's group
    }
    T getGenerator() {
       return generator;
    }
    T power(T a, unsigned long b) {
        T result = generator.one();
        while (b > 0) {
            if (b % 2 == 1) {
                result *= a;
            }
            a *= a;
            b /= 2;
        }
        return result;
    }
    unsigned long getRandomSecret() {
        std::random_device rd;
        std::mt19937 gen(rd());
        std::uniform_int_distribution dist(0, generator.getCharacteristic()-1);
        return dist(gen);
    }


};

File: jpp/lab/l4/z3/DHSetup.cs
using System;

namespace GFNamespace {
    public class DHSetup<T> where T : GF{
        public T generator { get; }

        public DHSetup(T element) {
            this.generator = ~element;
        }


        public T power(T a, long b) {
            T result = default(T);
            if (a.Equals(default(T))) {
                result = generator;
            } else {
                result = (dynamic)generator.one();
            }
            while (b > 0) {
              if (b % 2 == 1) {
                  result = (T)((dynamic)result * a);
              }
                a = (T)((dynamic)a * a);
                b /= 2;
              }
            return result;
        }

        public long getRandomSecret() {
            Random rand = new Random();
            return rand.Next(0, generator.characteristic);
        }
    }
}
